### ISSUE TITLE
[Q][archlinux] ckan git build master fails - why ?

### ISSUE THREAD

**Issue by birdspider** [KSP-CKAN/CKAN-support#49]
after some woes with git ssh and [ckan-git](https://aur.archlinux.org/packages/ckan-git/) on Archlinux User Repos, the user nlight/AUR pointed out that bin/ckan-build.py should to it.

Alas it is not so. Is the following simply a non compiling master or do I miss something during build ?

PS: I also made a weird little [ckan](https://aur.archlinux.org/packages/ckan/) package which in essence gets the latest tag (git pull, describe) and wgets the release from github. But I assume some users would prefer a working `ckan-git` option.
 
````
birdspider@blackbox ~/haha/CKAN-master$ python2 bin/ckan-build.py 
Building CKAN from the following commit hashes:                                                                                                                                     
CKAN-core: master
CKAN-GUI: master 
CKAN-cmdline: master            
Klone nach 'CKAN-core'... 

<snip git traffic output/>

+ xbuild /verbosity:minimal CKAN-core.sln
XBuild Engine Version 12.0
Mono, Version 3.10.0.0
Copyright (C) 2005-2013 Various Mono authors
                Configuration: Debug Platform: AnyCPU
FileIdentifier.cs(10,38): warning CS0414: The private field `CKAN.FileIdentifier.log' is assigned but its value is never used
Configuration: Debug Platform: AnyCPU
+ nunit-console --exclude=FlakyNetwork Tests/bin/Debug/Tests.dll
NUnit version 2.4.8
Copyright (C) 2002-2007 Charlie Poole.
Copyright (C) 2002-2004 James W. Newkirk, Michael C. Two, Alexei A. Vorontsov.
Copyright (C) 2000-2002 Philip Craig.
All Rights Reserved.

Runtime Environment - 
   OS Version: Unix 3.17.6.1
  CLR Version: 2.0.50727.1433 ( 3.10.0 (tarball Mon Oct  6 20:46:04 UTC 2014) )

Unhandled Exception:
System.TypeLoadException: Could not load type 'CKANTests.ModuleInstaller' from assembly 'Tests, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null'.

Server stack trace: 
  at (wrapper managed-to-native) System.MonoType:GetMethodsByName (System.MonoType,string,System.Reflection.BindingFlags,bool,System.Type)
  at System.MonoType.GetMethods (BindingFlags bindingAttr) [0x00000] in /build/mono/src/mono-3.10.0/mcs/class/corlib/System/MonoType.cs:230 
  at NUnit.Core.Reflect.GetMethodWithAttribute (System.Type fixtureType, System.String attributeName, BindingFlags bindingFlags, Boolean inherit) [0x00000] in /build/mono/src/mono-3.10.0/mcs/nunit24/NUnitCore/core/Reflect.cs:212 
  at NUnit.Core.NUnitFramework.GetSetUpMethod (System.Type fixtureType) [0x00000] in /build/mono/src/mono-3.10.0/mcs/nunit24/NUnitCore/core/NUnitFramework.cs:103 
  at NUnit.Core.NUnitTestMethod..ctor (System.Reflection.MethodInfo method) [0x00000] in /build/mono/src/mono-3.10.0/mcs/nunit24/NUnitCore/core/NUnitTestMethod.cs:17 
  at NUnit.Core.Builders.NUnitTestCaseBuilder.MakeTestCase (System.Reflection.MethodInfo method) [0x00000] in /build/mono/src/mono-3.10.0/mcs/nunit24/NUnitCore/core/Builders/NUnitTestCaseBuilder.cs:59 
  at NUnit.Core.Builders.AbstractTestCaseBuilder.BuildFrom (System.Reflection.MethodInfo method) [0x00013] in /build/mono/src/mono-3.10.0/mcs/nunit24/NUnitCore/core/Builders/AbstractTestCaseBuilder.cs:79 
  at NUnit.Core.Extensibility.TestCaseBuilderCollection.BuildFrom (System.Reflection.MethodInfo method) [0x00029] in /build/mono/src/mono-3.10.0/mcs/nunit24/NUnitCore/core/Extensibility/TestCaseBuilderCollection.cs:53 
  at NUnit.Core.Builders.AbstractFixtureBuilder.BuildTestCase (System.Reflection.MethodInfo method) [0x00028] in /build/mono/src/mono-3.10.0/mcs/nunit24/NUnitCore/core/Builders/AbstractFixtureBuilder.cs:174 
  at NUnit.Core.Builders.AbstractFixtureBuilder.AddTestCases (System.Type fixtureType) [0x00020] in /build/mono/src/mono-3.10.0/mcs/nunit24/NUnitCore/core/Builders/AbstractFixtureBuilder.cs:144 
  at NUnit.Core.Builders.AbstractFixtureBuilder.BuildFrom (System.Type type) [0x0001a] in /build/mono/src/mono-3.10.0/mcs/nunit24/NUnitCore/core/Builders/AbstractFixtureBuilder.cs:73 
  at NUnit.Core.Extensibility.SuiteBuilderCollection.BuildFrom (System.Type type) [0x00029] in /build/mono/src/mono-3.10.0/mcs/nunit24/NUnitCore/core/Extensibility/SuiteBuilderCollection.cs:55 
  at NUnit.Core.TestFixtureBuilder.BuildFrom (System.Type type) [0x00000] in /build/mono/src/mono-3.10.0/mcs/nunit24/NUnitCore/core/TestFixtureBuilder.cs:31 
  at NUnit.Core.Builders.TestAssemblyBuilder.GetFixtures (System.Reflection.Assembly assembly, System.String ns) [0x00032] in /build/mono/src/mono-3.10.0/mcs/nunit24/NUnitCore/core/Builders/TestAssemblyBuilder.cs:189 
  at NUnit.Core.Builders.TestAssemblyBuilder.Build (System.String assemblyName, Boolean autoSuites) [0x0001a] in /build/mono/src/mono-3.10.0/mcs/nunit24/NUnitCore/core/Builders/TestAssemblyBuilder.cs:99 
  at NUnit.Core.Builders.TestAssemblyBuilder.Build (System.String assemblyName, System.String testName, Boolean autoSuites) [0x00016] in /build/mono/src/mono-3.10.0/mcs/nunit24/NUnitCore/core/Builders/TestAssemblyBuilder.cs:76 
  at NUnit.Core.TestSuiteBuilder.BuildSingleAssembly (NUnit.Core.TestPackage package) [0x0001e] in /build/mono/src/mono-3.10.0/mcs/nunit24/NUnitCore/core/TestSuiteBuilder.cs:119 
  at NUnit.Core.TestSuiteBuilder.Build (NUnit.Core.TestPackage package) [0x00025] in /build/mono/src/mono-3.10.0/mcs/nunit24/NUnitCore/core/TestSuiteBuilder.cs:65 
  at NUnit.Core.SimpleTestRunner.Load (NUnit.Core.TestPackage package) [0x0000b] in /build/mono/src/mono-3.10.0/mcs/nunit24/NUnitCore/core/SimpleTestRunner.cs:107 
  at NUnit.Core.ProxyTestRunner.Load (NUnit.Core.TestPackage package) [0x00000] in /build/mono/src/mono-3.10.0/mcs/nunit24/NUnitCore/core/ProxyTestRunner.cs:104 
  at NUnit.Core.ProxyTestRunner.Load (NUnit.Core.TestPackage package) [0x00000] in /build/mono/src/mono-3.10.0/mcs/nunit24/NUnitCore/core/ProxyTestRunner.cs:104 
  at NUnit.Core.RemoteTestRunner.Load (NUnit.Core.TestPackage package) [0x0005b] in /build/mono/src/mono-3.10.0/mcs/nunit24/NUnitCore/core/RemoteTestRunner.cs:45 
  at (wrapper xdomain-dispatch) NUnit.Core.RemoteTestRunner:Load (object,byte[]&,byte[]&)

Exception rethrown at [0]: 

  at (wrapper xdomain-invoke) NUnit.Core.RemoteTestRunner:Load (NUnit.Core.TestPackage)
  at NUnit.Util.TestDomain.Load (NUnit.Core.TestPackage package) [0x0003f] in /build/mono/src/mono-3.10.0/mcs/nunit24/ClientUtilities/util/TestDomain.cs:57
````

**Comment by AlexanderDzhoganov** [comment]
This is weird, just tested on a fresh Ubuntu install with only mono-complete and it seems to compile.

**Comment by birdspider** [comment]
hmm, NUnit seems to be a testing framework, maybe its simply not installed - (no c#/mono knowledge  whatsoever)

**Comment by AlexanderDzhoganov** [comment]
It's downloaded as part of CKAN-core/packages

**Comment by birdspider** [comment]
could it be that `nunit-console` uses the systemwide nunit installation (see output: NUnit version 2.4.8) instead of the packaged one (`CKAN-core/packages/NUnit.2.6.3`) - and how would I change/test this?

**Comment by birdspider** [comment]
so after patching `bin/ckan_github_utils.py` during build I made __some progress__
right now this seems to be archlinux-specific so I refrain from proposing an upstream fix.

the patch `bin/ckan_github_utils.py@148`:
`if os.system('sh build.sh) != 0:` transforms to 
`if os.system('cat build.sh | sed "s/-console/-console4/m" | sh') != 0:`

I get the following __test fails__: 
Are they supposed to pass, should they be ignored (if so how ?)
_see how its using nunit-console4 !!_ that was the root problem
````
+ xbuild /verbosity:minimal CKAN-core.sln
XBuild Engine Version 12.0
Mono, Version 3.10.0.0
Copyright (C) 2005-2013 Various Mono authors
                Configuration: Debug Platform: AnyCPU
FileIdentifier.cs(10,38): warning CS0414: The private field `CKAN.FileIdentifier.log' is assigned but its value is never used
Configuration: Debug Platform: AnyCPU
+ nunit-console4 --nologo --exclude=FlakyNetwork Tests/bin/Debug/Tests.dll
Excluded categories: FlakyNetwork
.......................................F.F....F.F......./tmp/tmp32e3cc5.tmp/CKAN/downloads
...N.N.N.N.N........................N.N.........................N.N.N.N.N.N.N.N.N.N..N.N.N.N...N......
Tests run: 110, Failures: 4, Not run: 22, Time: 1.290 seconds

Test Case Failures:
1) CKANTests.Meta.ReleaseNumber : NUnit.Framework.InconclusiveException : No version detected - development build
at CKANTests.Meta.ReleaseNumber () [0x0000d] in /home/patrik/Projekte/XXX/src/CKAN/CKAN-core/Tests/CKAN/Meta.cs:30
at (wrapper managed-to-native) System.Reflection.MonoMethod:InternalInvoke (System.Reflection.MonoMethod,object,object[],System.Exception&)
at System.Reflection.MonoMethod.Invoke (System.Object obj, BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00054] in /build/mono/src/mono-3.10.0/mcs/class/corlib/System.Reflection/MonoMethod.cs:230

2) CKANTests.Meta.Version : NUnit.Framework.InconclusiveException : No version detected - development build
at CKANTests.Meta.Version () [0x00017] in /home/patrik/Projekte/XXX/src/CKAN/CKAN-core/Tests/CKAN/Meta.cs:15
at (wrapper managed-to-native) System.Reflection.MonoMethod:InternalInvoke (System.Reflection.MonoMethod,object,object[],System.Exception&)
at System.Reflection.MonoMethod.Invoke (System.Object obj, BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00054] in /build/mono/src/mono-3.10.0/mcs/class/corlib/System.Reflection/MonoMethod.cs:230

3) CKANTests.Module.FutureModule : NUnit.Framework.InconclusiveException : Dev build
at CKANTests.Module.FutureModule () [0x0000b] in /home/patrik/Projekte/XXX/src/CKAN/CKAN-core/Tests/CKAN/Types/Module.cs:104
at (wrapper managed-to-native) System.Reflection.MonoMethod:InternalInvoke (System.Reflection.MonoMethod,object,object[],System.Exception&)
at System.Reflection.MonoMethod.Invoke (System.Object obj, BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00054] in /build/mono/src/mono-3.10.0/mcs/class/corlib/System.Reflection/MonoMethod.cs:230

4) CKANTests.Module.IsSpecSupported : NUnit.Framework.InconclusiveException : Dev build
at CKANTests.Module.IsSpecSupported () [0x0000b] in /home/patrik/Projekte/XXX/src/CKAN/CKAN-core/Tests/CKAN/Types/Module.cs:77
at (wrapper managed-to-native) System.Reflection.MonoMethod:InternalInvoke (System.Reflection.MonoMethod,object,object[],System.Exception&)
at System.Reflection.MonoMethod.Invoke (System.Object obj, BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00054] in /build/mono/src/mono-3.10.0/mcs/class/corlib/System.Reflection/MonoMethod.cs:230
````

**Comment by birdspider** [comment]
> @AlexanderDzhoganov - NUnit gets downloaded as part of CKAN-core/packages

well yes, but I didn't see any downloaded `nunit-console` supplied so it.
So it then uses the systemwide installed one - which defaults to `mono 2.0` with `nunit 2.4.8` (at least on arch).

**Comment by birdspider** [comment]
well I give up, my suggestion is to update the various build scripts (build.sh) to have their nunit-console command optionally given by an ENV var. I have spent far to much time with this. 
In the process I needed to package nunit2.6 for arch (ubuntu seems to provide libnunit2.6-cil). 

Would I proceed the result would be a brittle `s/x/y/` which I tried in the python git script but then realized there is a build.sh calling a build.sh. Another reason is because from a packager perspective one cannot intervene (read patch) after the 3 gits get cloned because build immediatly starts.
